packages/teleport/src/DesktopSession/WarningDropdown.tsx

```diff
@@ -0,0 +1,148 @@
+/*
+Copyright 2022 Gravitational, Inc.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+import React, { useRef, useState, useEffect } from 'react';
+import { Text, Flex, Button, Card, ButtonIcon } from 'design';
+import styled from 'styled-components';
+import { Notification, NotificationItem } from 'shared/components/Notification';
+import { Warning, Close } from 'design/Icon';
+import { orange } from 'design/theme/palette';
+
+export function WarningDropdown(props: Props) {
+  const { warnings, onRemoveWarning } = props;
+
+  const [showDropdown, setShowDropdown] = useState(false);
+  const ref = useRef(null);
+
+  const toggleDropdown = () => {
+    if (warnings.length > 0) {
+      setShowDropdown(prevState => !prevState);
+    }
+  };
+
+  // Dropdown is always closed if there are no errors to show
+  useEffect(() => {
+    if (warnings.length === 0) setShowDropdown(false);
+  }, [warnings]);
```

Comment:
I think this falls under [the "Adjusting some state when a prop changes" section](https://beta.reactjs.org/learn/you-might-not-need-an-effect#adjusting-some-state-when-a-prop-changes) from You Might Not Need an Effect. There's also some advice on how to address this, I suppose in this case we could move this conditional outside of effect and it would just work.

Here the effect runs on every render anyway because `useEffect` uses `Object.is` to compare effect dependencies.

Reply:
> I think this falls under [the "Adjusting some state when a prop changes" section](https://beta.reactjs.org/learn/you-might-not-need-an-effect#adjusting-some-state-when-a-prop-changes) from You Might Not Need an Effect. There's also some advice on how to address this, I suppose in this case we could move this conditional outside of effect and it would just work.

👍

> Here the effect runs on every render anyway because `useEffect` uses `Object.is` to compare effect dependencies.

I don't understand why using `Object.is` means this runs on every render. Wouldn't it only run on renders induced by a call to `setWarnings`?

https://github.com/gravitational/webapps/blob/b22a8a7fa943e71e643e1ca95e1cb8e7f80d7114/packages/teleport/src/DesktopSession/useDesktopSession.tsx#L95

Reply:
Right, but won't it be precisely the same object unless I call `setWarnings`? Or is that not guaranteed by React's `useState`?

Reply:
> I don't understand why using `Object.is` means this runs on every render. Wouldn't it only run on renders induced by a call to `setWarnings`?

Oh sorry, I didn't follow how `warnings` is created back to its source. I think you might be right here.